### PR TITLE
[FSDP] Make recursive_wrap, wrap APIs independent of ConfigAutoWrap.

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69358
* #69357
* #69356
* __->__ #68776

Makes these APIs independent of ConfigAutoWrap so that they can be
used by FSDP ctor without it knowing about ConfigAutoWrap. Although this has been done in a way such that the APIs are fully backward compatible.

Also gets us one step closer to killing ConfigAutoWrap.recursive_wrap and
auto_wrap(), as we will only support enable_wrap() and wrap() moving forward.

ConfigAutoWrap.recursive_wrap and wrap() are refactored to reduce code duplication. 

Will test via unittests and FSDP benchmarks to ensure the wrapping still works.

Closes https://github.com/pytorch/pytorch/issues/68241

Differential Revision: [D32604021](https://our.internmc.facebook.com/intern/diff/D32604021/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/distributed/_fsdp/wrap.py [hunk 1]
```diff
@@ -161,16 +161,89 @@ def wrap(module: nn.Module, **wrap_overrides: Any) -> nn.Module:
...
+    # below logic by passing check_module_overrides=False.
+    if check_module_overrides and hasattr(module, "wrapper_config"):
+        module_overrides = module.wrapper_config
+        assert isinstance(module_overrides, dict)
+    wrap_overrides = {**kwargs, **module_overrides}
+    assert wrapper_cls is not None
+    return wrapper_cls(module, **wrap_overrides)
+
+
+def recursive_wrap(
```
Comment (zhaojuanmao) [inline torch/distributed/_fsdp/wrap.py:197]: nit: make it as private as well?

File: torch/distributed/_fsdp/wrap.py [hunk 2]
```diff
@@ -161,16 +161,89 @@ def wrap(module: nn.Module, **wrap_overrides: Any) -> nn.Module:
...
+        remainder = num_params - total_wrapped_params
+        if not only_wrap_children and auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder):
+            # Leaf node or final wrapping of the remainder both happen here.
+            return _wrap(module, wrapper_cls, **kwargs), num_params
+        else:
+            return module, total_wrapped_params
+    return module, 0
+
+
 def auto_wrap(module: nn.Module, auto_wrap_policy: Optional[Callable] = None, **kwargs: Any) -> nn.Module:
```
Comment (zhaojuanmao) [inline torch/distributed/_fsdp/wrap.py:247]: since no one is using this, we can just kill it?

File: torch/distributed/_fsdp/wrap.py [hunk 3]
```diff
@@ -257,7 +330,6 @@ def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
     def recursive_wrap(
```
Comment (zhaojuanmao) [inline torch/distributed/_fsdp/wrap.py:330]: kill it since no one is using it

File: torch/distributed/_fsdp/wrap.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/_fsdp/wrap.py:247]: I think right now our internal benchmarks are using it, I was going to remove it from those and then remove these APIs

**Comment by rohan-varma** [comment]
@zhaojuanmao Made the recursive_wrap private, planning to kill auto_wrap, ConfigAutoWrap.recursive_wrap in later diffs after removing them from our internal benchmarks.

File: torch/distributed/_fsdp/wrap.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/_fsdp/wrap.py:247]: benchmarks are not actually using it so i removed it in the above diffs

**Review by zhaojuanmao — APPROVED** [review]

**Pull request closed without merging**